dataservices/dataservices.go

```diff
@@ -0,0 +1,239 @@
+	//TODO: fill in the details
+	if permissions := r.Env[user.PERMISSIONS].(*user.UsersPermissions); permissions != nil {
+		return true
+	}
+	return false
+}
+
+//GetVersion will return the current API version
+func (client *DataServiceClient) GetVersion(w rest.ResponseWriter, r *rest.Request) {
+	w.WriteJson(version.Long())
+	return
+}
+
+//PostDataset will process a posted dataset for the requested user if permissons are sufficient
+func (client *DataServiceClient) PostDataset(w rest.ResponseWriter, r *rest.Request) {
+
+	userid := r.PathParam(useridParamName)
+
+	log.AddTrace(userid)
+
+	if checkPermisson(r, user.Permission{}) {
+
+		groupID := r.Env[user.GROUPID]
+
+		if r.ContentLength == 0 || groupID == "" {
+			rest.Error(w, missingDataError, http.StatusBadRequest)
+			return
+		}
+
+		var dataSet data.Dataset
+		var processedDataset struct {
+			Dataset []interface{} `json:"Dataset"`
+			Errors  string        `json:"Errors"`
+		}
+
+		err := r.DecodeJsonPayload(&dataSet)
+
+		if err != nil {
+			rest.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		platformData, err := data.NewTypeBuilder(map[string]interface{}{data.UserIDField: userid, data.GroupIDField: groupID}).BuildFromDataSet(dataSet)
+		processedDataset.Dataset = platformData
+		processedDataset.Errors = err.Error()
+
+		if err.Error() != "" {
+			w.WriteHeader(http.StatusBadRequest)
+			w.WriteJson(&processedDataset)
+			return
+		}
+
+		//TODO: should this be a bulk insert?
+		for i := range platformData {
+			if err = client.dataStore.Save(platformData[i]); err != nil {
+				rest.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		}
+
+		w.WriteJson(&processedDataset)
+		return
+	}
+	rest.Error(w, missingPermissionsError, http.StatusUnauthorized)
+	return
+}
+
+//PostBlob will save a posted blob of  data for the requested user if permissons are sufficient
+func (client *DataServiceClient) PostBlob(w rest.ResponseWriter, r *rest.Request) {
+
+	userid := r.PathParam(useridParamName)
+	log.AddTrace(userid)
+
+	if checkPermisson(r, user.Permission{}) {
+		w.WriteHeader(http.StatusNotImplemented)
+		return
+	}
+	rest.Error(w, missingPermissionsError, http.StatusUnauthorized)
+	return
+
+}
+
+//process the found data and send the appropriate response
+func process(iter store.Iterator) data.Dataset {
+
+	var chunk data.Datum
+	var all = data.Dataset{}
+
+	for iter.Next(&chunk) {
+		all = append(all, chunk)
+	}
+
+	return all
+}
+
+//GetDataset will return the requested users data set if permissons are sufficient
+func (client *DataServiceClient) GetDataset(w rest.ResponseWriter, r *rest.Request) {
+
+	log.AddTrace(r.PathParam(useridParamName))
+
+	if checkPermisson(r, user.Permission{}) {
+
+		var found struct {
+			data.Dataset `json:"Dataset"`
+			Errors       string `json:"Errors"`
+		}
+
+		userid := r.PathParam(useridParamName)
+		log.Info(useridParamName, userid)
+
+		types := strings.Split(r.URL.Query().Get("type"), ",")
```

Comment:
I'd call these variables `type, `subType`, `startDate`, and `endDate` to be more accurate.